src/sql/src/plan/with_options.rs

```diff
@@ -334,6 +334,33 @@ impl ImpliedValue for u32 {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Hash, Deserialize)]
+pub struct OptionalU32(pub Option<u32>);
+
+impl From<u32> for OptionalU32 {
+    fn from(i: u32) -> OptionalU32 {
+        OptionalU32(Some(i))
+    }
+}
+
+impl TryFromValue<Value> for OptionalU32 {
+    fn try_from_value(v: Value) -> Result<Self, PlanError> {
+        Ok(match v {
+            Value::Null => OptionalU32(None),
+            v => u32::try_from_value(v)?.into(),
+        })
+    }
+    fn name() -> String {
+        "optional uint4".to_string()
+    }
+}
+
+impl ImpliedValue for OptionalU32 {
+    fn implied_value() -> Result<Self, PlanError> {
+        sql_bail!("must provide an integer value or NULL")
+    }
+}
+
```

Comment:
Does implementing `impl TryFromValue<Value> for Option<u32>` and `impl ImpliedValue for Option<u32>` directly not work here?

Reply:
Good question; the main problem appears to be with `TryFromValue`. If we implement it for `Option<u32>`, then the compiler starts to believe that the following two implementation are conflicting: https://github.com/MaterializeInc/materialize/blob/5af439e9e8e8ef18f5a829b3966654a4a029b026/src/sql/src/plan/with_options.rs#L384-L432.

However, if we do not implement it for `Option<u32>`, then we do not get exactly the behavior we're looking for. In addition, the compiler is also unhappy within the macro `generate_extracted_config!`, complaining that it has trouble picking the correct `From` implementation.

Without much hope, I also tried to define this redundantly as `impl TryFromValue<WithOptionValue<u32>> for Option<u32> `. The compiler is unhappy again due to the conflicting implementation for `Option<V>`: https://github.com/MaterializeInc/materialize/blob/5af439e9e8e8ef18f5a829b3966654a4a029b026/src/sql/src/plan/with_options.rs#L384-L394.

Happy to try out any other suggestions, but looks like the `OptionalU32` pattern is pretty understandable and matches what we do for `OptionalInterval`?